File: app/src/main/java/wangdaye/com/geometricweather/common/ui/adapters/TagAdapter.java
package wangdaye.com.geometricweather.common.ui.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.ui.widgets.TagView;

public class TagAdapter extends RecyclerView.Adapter<TagAdapter.ViewHolder> {

    private final List<Tag> mTagList;
    private final @ColorInt int mTagTitleColor;
    private final @ColorInt int mCheckedBackgroundColor;
    private final @ColorInt int mUncheckedBackgroundColor;
    private final OnTagCheckedListener mListener;
    private int mCheckedIndex;

    public static final int UNCHECKABLE_INDEX = -1;

    class ViewHolder extends RecyclerView.ViewHolder {

        private final TagView mTagView;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            mTagView = itemView.findViewById(R.id.item_tag);
            mTagView.setOnClickListener(v -> {
                boolean consumed = false;
                if (mListener != null) {
                    consumed = mListener.onItemChecked(!mTagView.isChecked(), mCheckedIndex, getAdapterPosition());
                }
                if (!consumed && mCheckedIndex != getAdapterPosition()) {
                    int i = mCheckedIndex;
                    mCheckedIndex = getAdapterPosition();
                    notifyItemChanged(i);
                    notifyItemChanged(mCheckedIndex);
                }
            });
        }

        void onBindView(Tag tag, boolean checked) {
            mTagView.setText(tag.getName());
            mTagView.setTextColor(mTagTitleColor);
            setChecked(checked);

            mTagView.setCheckedBackgroundColor(mCheckedBackgroundColor);
            mTagView.setUncheckedBackgroundColor(mUncheckedBackgroundColor);
        }

        public void setChecked(boolean checked) {
            mTagView.setChecked(checked);
        }
    }

    public interface OnTagCheckedListener {
        boolean onItemChecked(boolean checked, int oldPosition, int newPosition);
    }

    public interface Tag {
        String getName();
    }

    public TagAdapter(List<Tag> tagList,
                      @ColorInt int tagTitleColor,
                      @ColorInt int checkedBackgroundColor,
                      @ColorInt int uncheckedBackgroundColor,
                      OnTagCheckedListener listener) {
        this(
                tagList,
                tagTitleColor,
                checkedBackgroundColor,
                uncheckedBackgroundColor,
                listener,
                UNCHECKABLE_INDEX
        );
    }

    public TagAdapter(List<Tag> tagList,
                      @ColorInt int tagTitleColor,
                      @ColorInt int checkedBackgroundColor,
                      @ColorInt int uncheckedBackgroundColor,
                      OnTagCheckedListener listener,
                      int checkedIndex) {
        mTagList = tagList;
        mTagTitleColor = tagTitleColor;
        mCheckedBackgroundColor = checkedBackgroundColor;
        mUncheckedBackgroundColor = uncheckedBackgroundColor;
        mListener = listener;
        mCheckedIndex = checkedIndex;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_tag, parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.onBindView(mTagList.get(position), position == mCheckedIndex);
    }

    @Override
    public int getItemCount() {
        return mTagList.size();
    }

    public void insertItem(Tag tag) {
        mTagList.add(tag);
        notifyItemInserted(mTagList.size() - 1);
    }

    public Tag removeItem(int position) {
        Tag tag = mTagList.remove(position);
        notifyItemRemoved(position);
        return tag;
    }
}

File: app/src/main/java/org/breezyweather/common/ui/adapters/TagAdapter.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import androidx.recyclerview.widget.RecyclerView
import org.breezyweather.R
import org.breezyweather.common.ui.widgets.TagView

class TagAdapter @JvmOverloads constructor(
    private val mTagList: MutableList<Tag>,
    @field:ColorInt @param:ColorInt private val mCheckedTitleColor: Int,
    @field:ColorInt @param:ColorInt private val mUncheckedTitleColor: Int,
    @field:ColorInt @param:ColorInt private val mCheckedBackgroundColor: Int,
    @field:ColorInt @param:ColorInt private val mUncheckedBackgroundColor: Int,
    private val mListener: ((checked: Boolean, oldPosition: Int, newPosition: Int) -> Boolean)? = null,
    private var mCheckedIndex: Int = UNCHECKABLE_INDEX,
) : RecyclerView.Adapter<TagAdapter.ViewHolder>() {
    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val mTagView: TagView

        init {
            mTagView = itemView.findViewById(R.id.item_tag)
            mTagView.setOnClickListener {
                var consumed = false
                if (mListener != null) {
                    consumed = mListener.invoke(
                        !mTagView.isChecked,
                        mCheckedIndex,
                        bindingAdapterPosition
                    )
                }
                if (!consumed && mCheckedIndex != bindingAdapterPosition) {
                    val i = mCheckedIndex
                    mCheckedIndex = bindingAdapterPosition
                    notifyItemChanged(i)
                    notifyItemChanged(mCheckedIndex)
                }
            }
        }

        fun onBindView(tag: Tag, checked: Boolean) {
            mTagView.apply {
                text = tag.name
                checkedBackgroundColor = mCheckedBackgroundColor
                uncheckedBackgroundColor = mUncheckedBackgroundColor
            }
            setChecked(checked)
        }

        fun setChecked(checked: Boolean) {
            mTagView.apply {
                setTextColor(if (checked) mCheckedTitleColor else mUncheckedTitleColor)
                isChecked = checked
            }
        }
    }

    interface Tag {
        val name: String
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_tag, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.onBindView(mTagList[position], position == mCheckedIndex)
    }

    override fun getItemCount(): Int {
        return mTagList.size
    }

    fun insertItem(tag: Tag) {
        mTagList.add(tag)
        notifyItemInserted(mTagList.size - 1)
    }

    fun removeItem(position: Int): Tag {
        val tag = mTagList.removeAt(position)
        notifyItemRemoved(position)
        return tag
    }

    companion object {
        const val UNCHECKABLE_INDEX = -1
    }
}
